url29ff2e53f3_nicolas3470_Settlers_of_Catan_AI_tgz-pJ8-robotGameJ8/src/soc/game/SOCGame.java

```diff
@@ -0,0 +1,4792 @@
+        {
+            // No one is left to special build.
+            // Case "A" or "D".
+            if (alreadyInPhase)
+            {
+                currentPlayerNumber = specialBuildPhase_afterPlayerNumber;
+                gameState = PLAY1;  // case "D"
+            }
+            return false;
+        }
+
+        // Case "B" or "C".
+        if (! alreadyInPhase)
+        {
+            // case "B":
+
+            // usually the current player can't call for SBP.
+            // unless, that is, it's the very start of their turn
+            // and they haven't rolled or played a card yet.
+
+            if (players[prevPlayer].hasAskedSpecialBuild()
+                && (gameState == PLAY)
+                && ! players[prevPlayer].hasPlayedDevCard())
+            {
+                // remember previous player, re-set current player:
+
+                gameState = SPECIAL_BUILDING;
+                currentPlayerNumber = prevPlayer;
+                if (! advanceTurnBackwards())
+                    return false;  // all players have left
+                specialBuildPhase_afterPlayerNumber = currentPlayerNumber;
+                currentPlayerNumber = prevPlayer;
+            } else {
+                // usual case: ending current player's turn.
+
+                gameState = SPECIAL_BUILDING;
+                specialBuildPhase_afterPlayerNumber = prevPlayer;
+            }
+        }
+        // currentPlayerNumber is set already by advanceTurn.
+
+        return true;
+    }
+
+    /**
+     * Put this piece on the board and update all related game state.
+     * May change current player and gamestate.
+     *<P>
+     * For example, if game state when called is {@link #START2A},
+     * this is their second initial settlement, so give
+     * the player some resources, and call their {@link SOCPlayer#clearPotentialSettlements()}.
+     *<P>
+     * Calls {@link SOCBoard#putPiece(SOCPlayingPiece)} and {@link SOCPlayer#putPiece(SOCPlayingPiece)}.
+     * Updates longest road if necessary.
+     * Calls {@link #advanceTurnStateAfterPutPiece()}.
+     *<P>
+     * If the piece is a city, putPiece removes the settlement there.
+     *
+     * @param pp the piece to put on the board
+     */
+    public void putPiece(SOCPlayingPiece pp)
+    {
+        /**
+         * call putPiece() on every player so that each
+         * player's updatePotentials() function gets called
+         */
+        for (int i = 0; i < maxPlayers; i++)
+        {
+            players[i].putPiece(pp);
+        }
+
+        board.putPiece(pp);
+
+        /**
+         * if the piece is a city, remove the settlement there
+         */
+        if (pp.getType() == SOCPlayingPiece.CITY)
+        {
+            SOCSettlement se = new SOCSettlement(pp.getPlayer(), pp.getCoordinates(), board);
+
+            for (int i = 0; i < maxPlayers; i++)
+            {
+                players[i].removePiece(se);
+            }
+
+            board.removePiece(se);
+        }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 1729:1739, 1890:1900, 2118:2128.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.